### ISSUE TITLE
Federation cluster scripts accidentally delete PV's.

### ISSUE THREAD

**Issue by madhusudancs** [kubernetes/kubernetes#46380]
We have a PVC that uses the alpha annotation to dynamically provision a GCE PD/PV. Sometimes in our test environment, the dynamically provisioned PV gets automatically deleted by the PV controller without any of us deleting the PVC. I am attaching the controller manager logs here: [kube-controller-manager.log-20170523-1495580401.gz](https://github.com/kubernetes/kubernetes/files/1026899/kube-controller-manager.log-20170523-1495580401.gz)

Interesting bits in the logs start at:

```
I0523 22:08:01.117123       5 gce_util.go:122] Successfully created GCE PD volume jenkins-us-central1-f--pvc-4723f163-4004-11e7-a75a-42010a80000a
```

The namespaced name of the claim is `f8n-system-agent-pr-93-0/e2e-f8n-agent-pr-93-0-apiserver-etcd-claim`

Here is the deployment yaml

```yaml
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  annotations:
    federation.alpha.kubernetes.io/federation-name: e2e-f8n-agent-pr-93-0
  labels:
    app: federated-cluster
  name: e2e-f8n-agent-pr-93-0-apiserver
  namespace: f8n-system-agent-pr-93-0
spec:
  replicas: 1
  selector:
    matchLabels:
      app: federated-cluster
      module: federation-apiserver
  strategy:
    rollingUpdate:
      maxSurge: 1
      maxUnavailable: 1
    type: RollingUpdate
  template:
    metadata:
      annotations:
        federation.alpha.kubernetes.io/federation-name: e2e-f8n-agent-pr-93-0
      creationTimestamp: null
      labels:
        app: federated-cluster
        module: federation-apiserver
      name: e2e-f8n-agent-pr-93-0-apiserver
    spec:
      containers:
      - command:
        - /hyperkube
        - federation-apiserver
        - --admission-control=NamespaceLifecycle
        - --advertise-address=104.154.166.121
        - --basic-auth-file=/etc/federation/apiserver/basicauth.csv
        - --bind-address=0.0.0.0
        - --client-ca-file=/etc/federation/apiserver/ca.crt
        - --etcd-servers=http://localhost:2379
        - --secure-port=8443
        - --tls-cert-file=/etc/federation/apiserver/server.crt
        - --tls-private-key-file=/etc/federation/apiserver/server.key
        - --token-auth-file=/etc/federation/apiserver/token.csv
        - --v=4
        image: gcr.io/k8s-jkns-pr-bldr-e2e-gce-fdrtn/hyperkube-amd64:v1.7.0-alpha.4.362_c5319821fe72fa
        imagePullPolicy: IfNotPresent
        name: apiserver
        ports:
        - containerPort: 8443
          name: https
          protocol: TCP
        - containerPort: 8080
          name: local
          protocol: TCP
        resources: {}
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /etc/federation/apiserver
          name: e2e-f8n-agent-pr-93-0-apiserver-credentials
          readOnly: true
      - command:
        - /usr/local/bin/etcd
        - --data-dir
        - /var/etcd/data
        image: gcr.io/google_containers/etcd:3.0.17
        imagePullPolicy: IfNotPresent
        name: etcd
        resources: {}
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /var/etcd
          name: etcddata
      dnsPolicy: ClusterFirst
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      terminationGracePeriodSeconds: 30
      volumes:
      - name: e2e-f8n-agent-pr-93-0-apiserver-credentials
        secret:
          defaultMode: 420
          secretName: e2e-f8n-agent-pr-93-0-apiserver-credentials
      - name: etcddata
        persistentVolumeClaim:
          claimName: e2e-f8n-agent-pr-93-0-apiserver-etcd-claim
```

cc @kubernetes/sig-storage-bugs @saad-ali @kubernetes/sig-federation-bugs

**Comment by madhusudancs** [comment]
PVC

```yaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  annotations:
    federation.alpha.kubernetes.io/federation-name: e2e-f8n-agent-pr-93-0
    volume.alpha.kubernetes.io/storage-class: "yes"
  labels:
    app: federated-cluster
  name: e2e-f8n-agent-pr-93-0-apiserver-etcd-claim
  namespace: e2e-f8n-agent-pr-93-0
spec:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 10Gi
```

**Comment by wongma7** [comment]
The PV got deleted somehow and so PVC entered Lost state. If we rule out that PV was accidentally deleted or it wasn't deleted at all then potentially there's an issue of the pv controller's cache getting out of sync with reality or something of that nature

I0523 22:08:01.361501       5 pv_controller_base.go:215] volume "pvc-4723f163-4004-11e7-a75a-42010a80000a" deleted

**Comment by wongma7** [comment]
I don't see "doDeleteVolume [pvc-4723f163-4004-11e7-a75a-42010a80000a" so I'm inclined to think the PV object was accidentally deleted by a user.

ping @jsafrane since this involves PV controller

**Comment by msau42** [comment]
@madhusudancs can you also attach apiserver logs?

**Comment by madhusudancs** [comment]
@msau42 we don't have the API server logs for the attached controller manager logs unfortunately. That cluster has been torn down. I can attach the API server logs belonging to a different instance when this happens again.

**Comment by jsafrane** [comment]
@wongma7 is right, it seems that something else than the PV controller deleted PV with name `pvc-4723f163-4004-11e7-a75a-42010a80000a`.
If the controller deleted the volume, we would see at least:
* `deleteVolumeOperation [%s] started` from PV controller 
* `Successfully deleted GCE PD volume %s` from volume plugin (at level 2!)
*  `volume %q deleted` from the controller again (at level 2!)

Looking at the log, there is quite lot of Lost PVCs... Some of the corresponding PVs are deleted very quickly after they're provisioned, some of them survived for couple of minutes. They are deleted in batches. Here is log from the controller where it provisioned 3 PVs during 3 minutes and they were deleted at the same time by something at the same time:

```
I0523 22:28:51.909294       5 pv_controller.go:1414] volume "pvc-30852a65-4007-11e7-a75a-42010a80000a" provisioned for claim "f8n-system-agent-pr-33-0/e2e-f8n-agent-pr-33-0-apiserver-etcd-claim"
I0523 22:29:35.778848       5 pv_controller.go:1414] volume "pvc-4aab8a3b-4007-11e7-a75a-42010a80000a" provisioned for claim "f8n-system-agent-pr-39-0/e2e-f8n-agent-pr-39-0-apiserver-etcd-claim"
I0523 22:30:28.335114       5 pv_controller.go:1414] volume "pvc-69f8571d-4007-11e7-a75a-42010a80000a" provisioned for claim "f8n-system-agent-pr-13-0/e2e-f8n-agent-pr-13-0-apiserver-etcd-claim"
I0523 22:31:01.482932       5 pv_controller.go:653] claim "f8n-system-agent-pr-33-0/e2e-f8n-agent-pr-33-0-apiserver-etcd-claim" entered phase "Lost"
I0523 22:31:01.505094       5 pv_controller.go:653] claim "f8n-system-agent-pr-39-0/e2e-f8n-agent-pr-39-0-apiserver-etcd-claim" entered phase "Lost"
I0523 22:31:01.522478       5 pv_controller.go:653] claim "f8n-system-agent-pr-13-0/e2e-f8n-agent-pr-13-0-apiserver-etcd-claim" entered phase "Lost"
```

Something must be watching the PVs periodically and deleting them. Do you have any 3rd party controllers / provisioners? Do you accidentally run second controller-manager in parallel? Is the GCE PD deleted too or just it's Kubernetes PV?

Watching the API server logs could help, especially if you could tell who deleted the PV object.

**Comment by madhusudancs** [comment]
@jsafrane @wongma7 you were right. We were running `kubectl delete pvc,pv,pods,... --namespace=${FEDERATION_NAME} --all` without paying close attention to what was going on there. Fix sent in #46945.

**Comment by saad-ali** [comment]
Good debugging @jsafrane!

**Comment by marun** [comment]
@madhusudancs Is there more to do with this issue that wasn't addressed by #46945?

**Comment by quinton-hoole** [comment]
Renamed and closed, as #46945 has merged.
